proxy/coco/src/state.rs

```diff
@@ -0,0 +1,733 @@
+/// High-level interface to the coco monorepo and gossip layer.
+pub struct State {
+    /// Internal handle on [`PeerApi`].
+    pub(crate) api: PeerApi<keys::SecretKey>,
+}
+
+impl State {
+    /// Create a new [`State`] given a [`PeerApi`].
+    #[must_use]
+    pub fn new(api: PeerApi<keys::SecretKey>, signer: signer::BoxedSigner) -> Self {
+        let paths = api.paths();
+
+        // Turns out we don't need to call the git commands from the command line.
+        // All we need is to register the git transport and this is the same
+        // thing. This means we can use git2 to execute actions. So instead of
+        // implementing the checkout via git clone as a Command, we use
+        // [`git2::RepoBuilder`].
+        transport::register(transport::Settings {
+            paths: paths.clone(),
+            signer,
+        });
+
+        Self { api }
+    }
+
+    /// Returns the [`PathBuf`] to the underlying monorepo.
+    #[must_use]
+    pub fn monorepo(&self) -> PathBuf {
+        self.api.paths().git_dir().join("")
+    }
+
+    /// Returns the underlying [`paths::Paths`].
+    #[must_use]
+    pub fn paths(&self) -> paths::Paths {
+        self.api.paths().clone()
+    }
+
+    /// Convenience method to trigger a reopen of the storage.
+    ///
+    /// # Errors
+    ///
+    /// When the underlying lock acquisition fails or opening the storage.
+    pub fn reopen(&self) -> Result<(), Error> {
+        self.api.storage().reopen()?;
+        Ok(())
+    }
+
+    /// Check the storage to see if we have the given commit for project at `urn`.
+    ///
+    /// # Errors
+    ///
+    ///   * Checking the storage for the commit fails.
+    ///
+    /// # Panics
+    ///
+    ///   * Unable to acquire the lock.
+    pub fn has_commit(&self, urn: &RadUrn, oid: impl Into<git2::Oid>) -> Result<bool, Error> {
+        Ok(self.api.storage().has_commit(urn, oid.into())?)
+    }
+
+    /// Our current peers [`PeerId`].
+    #[must_use]
+    pub fn peer_id(&self) -> PeerId {
+        self.api.peer_id().clone()
+    }
+
+    /// The address this peer is listening on.
+    #[must_use]
+    pub fn listen_addr(&self) -> SocketAddr {
+        self.api.listen_addr()
+    }
+
+    /// Get the default owner for this `PeerApi`.
+    #[must_use]
+    pub fn default_owner(&self) -> Option<user::User<entity::Draft>> {
+        match self.api.storage().default_rad_self() {
+            Ok(user) => Some(user),
+            Err(err) => {
+                log::warn!("an error occurred while trying to get 'rad/self': {}", err);
+                None
+            },
+        }
+    }
+
+    /// Set the default owner for this `PeerApi`.
+    ///
+    /// # Errors
+    ///
+    ///   * Fails to set the default `rad/self` for this `PeerApi`.
+    pub fn set_default_owner(&self, user: User) -> Result<(), Error> {
+        Ok(self.api.storage().set_default_rad_self(user)?)
+    }
+
+    /// Initialise a [`User`] and make them the default owner of this `PeerApi`.
```

Comment:
What do you mean by redundancy?

There can be only one owner, and setting the default owner lets us know which user entity that is -- because there will be multiple users in the monorepo which are other peer's users.

Reply:
@FintanH Can probably answer that.